### PR TITLE
Coverage test

### PR DESCRIPTION
**Pull request by BillyWooo**
- Add cargo tarpaulin 
- Add codecov upload
- Currently only cover part of the code space

### CODE REVIEW

File: .codecov.yml [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    project:
+      default:
+        threshold: 5%            # allows coverage to drop by up to the percent noted and still post a success status
+        #informational: true     # If true is specified the resulting status will pass no matter what the coverage is or what other settings are specified
+    patch: off
+
+ignore:
+  - "node"  # ignore folders and all its contents
+  - "runtime"  # ignore folders and all its contents
+  - "inspect"  # ignore folders and all its contents
```
Comment (Kailai-Wang) [inline .codecov.yml]: we don't have this folder (actually what does this file do?)

File: .github/workflows/coverage.yml [hunk 2]
```diff
@@ -0,0 +1,66 @@
+name: Tests Coverage
+
+on:
+  push:
+    branches:
+      - coverage_test
```
Comment (Kailai-Wang) [inline .github/workflows/coverage.yml]: should be `dev`

File: .github/workflows/coverage.yml [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+      - name: Install additional dependencies
+        run: |
+            sudo apt update -y &&
+            sudo apt install -y cmake pkg-config libssl-dev git gcc build-essential git clang libclang-dev
+
+      - name: Install cargo-tarpaulin
+        run: |
+            cargo install cargo-tarpaulin
+          # wget https://github.com/xd009642/tarpaulin/releases/download/${{ env.TARPAULIN_VERSION }}/cargo-tarpaulin-${{ env.TARPAULIN_VERSION }}-travis.tar.gz
+          # tar -zxvf cargo-tarpaulin-${{ env.TARPAULIN_VERSION }}-travis.tar.gz -C $HOME/.cargo/bin
```
Comment (Kailai-Wang) [inline .github/workflows/coverage.yml]: we can remove unused commands

File: .github/workflows/coverage.yml [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+      - name: Run cargo-tarpaulin
+        run: |
+          # cargo tarpaulin --release --no-fail-fast --workspace -e node primitives runtime scritps templates ts-tests
```
Comment (Kailai-Wang) [inline .github/workflows/coverage.yml]: we can remove unused commands

File: .github/workflows/coverage.yml [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    paths-ignore:
+      - "**/dependabot.yml"
+      - "**/README.md"
+  pull_request:
+    branches:
+      - dev
+  workflow_dispatch:
+
+env:
+  TARPAULIN_VERSION: 0.20.1
```
Comment (Kailai-Wang) [inline .github/workflows/coverage.yml]: - not used
- we can remove unused commands

**Review by Kailai-Wang — CHANGES_REQUESTED** [review]

File: .codecov.yml (hunk 1 above)
Comment (BillyWooo) [reply .codecov.yml]: Removed. This is copied from Acala for test purpose.

File: .github/workflows/coverage.yml (hunk 3 above)
Comment (BillyWooo) [reply .github/workflows/coverage.yml]: removed.

File: .github/workflows/coverage.yml (hunk 4 above)
Comment (BillyWooo) [reply .github/workflows/coverage.yml]: removed.

File: .github/workflows/coverage.yml (hunk 5 above)
Comment (BillyWooo) [reply .github/workflows/coverage.yml]: removed.

File: .github/workflows/coverage.yml (hunk 2 above)
Comment (BillyWooo) [reply .github/workflows/coverage.yml]: changed.

**Review by Kailai-Wang — APPROVED** [review]
Generally looks good to me, but there are still some things to tweak:
- [the files](https://codecov.io/gh/litentry/litentry-parachain/tree/4c4a62bbf8012121d408a55baa1998d2a34a2988) that are coverage-tested shouldn't include `node/` or `runtime/` (maybe we need to either explicitly exclude them or  use that .codecov config file)
- the benchmarking.rs and weights.rs should be excluded from the pallets (e.g. https://codecov.io/gh/litentry/litentry-parachain/tree/4c4a62bbf8012121d408a55baa1998d2a34a2988/pallets/drop3/src)

**Comment by BillyWooo** [comment]
> Generally looks good to me, but there are still some things to tweak:
> 
> * [the files](https://codecov.io/gh/litentry/litentry-parachain/tree/4c4a62bbf8012121d408a55baa1998d2a34a2988) that are coverage-tested shouldn't include `node/` or `runtime/` (maybe we need to either explicitly exclude them or  use that .codecov config file)
> * the benchmarking.rs and weights.rs should be excluded from the pallets (e.g. https://codecov.io/gh/litentry/litentry-parachain/tree/4c4a62bbf8012121d408a55baa1998d2a34a2988/pallets/drop3/src)

Yes. I agree. That makes sense.

**Comment by BillyWooo** [comment]
https://codecov.io/github/litentry/litentry-parachain/commit/1976048b10a79e6ff809bf9bada35126ac6a1f8a

**Pull request merged**